framework/di/Container.php

```diff
@@ -524,7 +525,7 @@ public function resolveCallableDependencies(callable $callback, $params = [])
                     unset($params[$name]);
                 } elseif (!$associative && isset($params[0]) && $params[0] instanceof $className) {
                     $args[] = array_shift($params);
-                } elseif (Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
+                } elseif (Yii::$app instanceof ServiceLocator && Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
```

Comment:
First of all: you still have not answered my first question:

> Does this check really needed?

Why extra check has been introduced at the first place? We were fine without it for long enogh. Why it should appear now?

> I strongly feel that in this case the argument for bad code in favor of perceived performance is not valid.

There is a class contract:
https://github.com/yiisoft/yii2/blob/master/framework/BaseYii.php#L71

It says that `Yii::$app` can be `Application` instance or it can be `null` - nothing else.
You don't use `instanceof` operator to check if `ActiveRecord::findOne()` actually returns `ActiveRecord` instance, do you?
The check you perform here is like following:

```php
$item = Item::findOne($id);
if (!$item instanceof Item) {
    ...
}
...
```

Instead simplier and faster solution used over all framework:

```php
$item = Item::findOne($id);
if ($item === null) {
    ...
}
...
```